File: kotlin/src/main/kotlin/com/github/lusing/qimen/ShiShen.kt
package com.github.lusing.qimen

open class ShiShen() : LiuQin() {
    var shiShen = 0
        protected set get

    override fun toString(): String {
        val sb = StringBuilder()
        when (shiShen) {
            ShiShen.Companion.ZHENGYIN -> sb.append("正印（印绶）")
            ShiShen.Companion.PIANYIN -> sb.append("偏印（枭）")
            ShiShen.Companion.SHANGGUAN -> sb.append("伤官")
            ShiShen.Companion.SHISHEN -> sb.append("食神")
            ShiShen.Companion.ZHENGGUAN -> sb.append("正官")
            ShiShen.Companion.QISHA -> sb.append("七杀(偏官)")
            ShiShen.Companion.ZHENGCAI -> sb.append("正财")
            ShiShen.Companion.PIANCAI -> sb.append("偏财")
            ShiShen.Companion.BIJIAN -> sb.append("比肩")
            ShiShen.Companion.JIECAI -> sb.append("劫财")
            else -> {}
        }
        return sb.toString()
    }

    companion object {
        /**
         * 生我者为父母 / 印星 异性为正印 同性为偏印
         */
        const val YINXING = 0x1
        const val ZHENGYIN = 0x10
        const val PIANYIN = 0x11

        /**
         * 我生者为儿女 / 食伤 异性为伤官 同性为食神
         */
        const val SHISHANG = 0x2
        const val SHANGGUAN = 0x20
        const val SHISHEN = 0x21

        /**
         * 克我者为官杀 异性为正官 同性为七杀
         */
        const val ZHENGGUAN = 0x30
        const val QISHA = 0x31
        val PIANGUAN: Int = ShiShen.Companion.QISHA

        /**
         * 我克者为妻财 / 财星 异性为正财 同性为偏财
         */
        const val CAIXING = 0x4
        const val ZHENGCAI = 0x40
        const val PIANCAI = 0x41

        /**
         * 同我者为兄弟 / 比劫 异性为劫财 同性为比肩
         */
        const val BIJIE = 0x05
        const val JIECAI = 0x50
        const val BIJIAN = 0x51
        fun getShiShen(me: TianGan, other: TianGan): ShiShen {
            val ss1 = ShiShen()
            val sum = me.tianGan + other.tianGan
            val isSame = sum % 2 == 0
            if (other.isSheng(me)) {
                ss1.liuqin = FUMU
                if (isSame) {
                    ss1.shiShen = ShiShen.Companion.ZHENGYIN
                } else {
                    ss1.shiShen = ShiShen.Companion.PIANYIN
                }
            } else if (me.isSheng(other)) {
                ss1.liuqin = ERNV
                if (isSame) {
                    ss1.shiShen = ShiShen.Companion.SHISHEN
                } else {
                    ss1.shiShen = ShiShen.Companion.SHANGGUAN
                }
            } else if (other.isKe(me)) {
                ss1.liuqin = GUANSHA
                if (isSame) {
                    ss1.shiShen = ShiShen.Companion.QISHA
                } else {
                    ss1.shiShen = ShiShen.Companion.ZHENGGUAN
                }
            } else if (me.isKe(other)) {
                ss1.liuqin = QICAI
                if (isSame) {
                    ss1.shiShen = ShiShen.Companion.PIANCAI
                } else {
                    ss1.shiShen = ShiShen.Companion.ZHENGCAI
                }
            } else {
                ss1.liuqin = XIONGDI
                if (isSame) {
                    ss1.shiShen = ShiShen.Companion.BIJIAN
                } else {
                    ss1.shiShen = ShiShen.Companion.JIECAI
                }
            }
            return ss1
        }
    }
}

File: js/shishen.js
const LiuQin = require('./liuqin.js');

class ShiShen extends LiuQin {
    constructor() {
        super(LiuQin.XIONGDI);
        this.shiShen = 0;
    }

    toLongString() {
        const sb = [];
        switch (this.shiShen) {
            case ShiShen.ZHENGYIN:
                sb.push("正印（印绶）");
                break;
            case ShiShen.PIANYIN:
                sb.push("偏印（枭）");
                break;
            case ShiShen.SHANGGUAN:
                sb.push("伤官");
                break;
            case ShiShen.SHISHEN:
                sb.push("食神");
                break;
            case ShiShen.ZHENGGUAN:
                sb.push("正官");
                break;
            case ShiShen.QISHA:
                sb.push("七杀(偏官)");
                break;
            case ShiShen.ZHENGCAI:
                sb.push("正财");
                break;
            case ShiShen.PIANCAI:
                sb.push("偏财");
                break;
            case ShiShen.BIJIAN:
                sb.push("比肩");
                break;
            case ShiShen.JIECAI:
                sb.push("劫财");
                break;
            default:
                break;
        }
        return sb.join('');
    }

    toString() {
        const sb = [];
        switch (this.shiShen) {
            case ShiShen.ZHENGYIN:
                sb.push("印");
                break;
            case ShiShen.PIANYIN:
                sb.push("枭");
                break;
            case ShiShen.SHANGGUAN:
                sb.push("伤");
                break;
            case ShiShen.SHISHEN:
                sb.push("食");
                break;
            case ShiShen.ZHENGGUAN:
                sb.push("官");
                break;
            case ShiShen.QISHA:
                sb.push("杀");
                break;
            case ShiShen.ZHENGCAI:
                sb.push("财");
                break;
            case ShiShen.PIANCAI:
                sb.push("才");
                break;
            case ShiShen.BIJIAN:
                sb.push("比");
                break;
            case ShiShen.JIECAI:
                sb.push("劫");
                break;
            default:
                break;
        }
        return sb.join('');
    }

    static get YINXING() {
        return 0x1;
    }

    static get ZHENGYIN() {
        return 0x10;
    }

    static get PIANYIN() {
        return 0x11;
    }

    static get SHISHANG() {
        return 0x2;
    }

    static get SHANGGUAN() {
        return 0x20;
    }

    static get SHISHEN() {
        return 0x21;
    }

    static get ZHENGGUAN() {
        return 0x30;
    }

    static get QISHA() {
        return 0x31;
    }

    static get PIANGUAN() {
        return ShiShen.QISHA;
    }

    static get CAIXING() {
        return 0x4;
    }

    static get ZHENGCAI() {
        return 0x40;
    }

    static get PIANCAI() {
        return 0x41;
    }

    static get BIJIE() {
        return 0x05;
    }

    static get JIECAI() {
        return 0x50;
    }

    static get BIJIAN() {
        return 0x51;
    }

    static getShiShen(me, other) {
        const ss1 = new ShiShen();
        const sum = me.tianGan + other.tianGan;
        const isSame = sum % 2 === 0;
        if (other.isSheng(me)) {
            ss1.liuqin = LiuQin.FUMU;
            ss1.shiShen = isSame ? ShiShen.PIANYIN : ShiShen.ZHENGYIN;
        } else if (me.isSheng(other)) {
            ss1.liuqin = LiuQin.ZISUN;
            ss1.shiShen = isSame ? ShiShen.SHISHEN : ShiShen.SHANGGUAN;
        } else if (other.isKe(me)) {
            ss1.liuqin = LiuQin.GUANGUI;
            ss1.shiShen = isSame ? ShiShen.QISHA : ShiShen.ZHENGGUAN;
        } else if (me.isKe(other)) {
            ss1.liuqin = LiuQin.QICAI;
            ss1.shiShen = isSame ? ShiShen.PIANCAI : ShiShen.ZHENGCAI;
        } else {
            ss1.liuqin = LiuQin.XIONGDI;
            ss1.shiShen = isSame ? ShiShen.BIJIAN : ShiShen.JIECAI;
        }
        return ss1;
    }
}

module.exports = ShiShen;
